src/Dialect/ONNX/ONNXOps/Math/RandomNormalLike.cpp

```diff
@@ -23,92 +24,11 @@ using namespace onnx_mlir;
-      return emitOpError("dtype not float16, float, double or bfloat16");
-    }
-    if (elementTypeID ==
-            onnx::TensorProto_DataType::TensorProto_DataType_FLOAT16 &&
-        outputType != Float16Type::get(getContext()))
-      return emitOpError("output tensor does not match float16 dtype.");
-    else if (elementTypeID ==
-                 onnx::TensorProto_DataType::TensorProto_DataType_FLOAT &&
-             outputType != Float32Type::get(getContext()))
-      return emitOpError("output tensor does not match float dtype.");
-    else if (elementTypeID ==
-                 onnx::TensorProto_DataType::TensorProto_DataType_DOUBLE &&
-             outputType != Float64Type::get(getContext()))
-      return emitOpError("output tensor does not match double dtype.");
-    else if (elementTypeID ==
-                 onnx::TensorProto_DataType::TensorProto_DataType_BFLOAT16 &&
-             outputType != BFloat16Type::get(getContext()))
-      return emitOpError("output tensor does not match bfloat16 dtype.");
-  } else if (inputType != outputType) {
-    return emitOpError("output and input element types do not match.");
-  }
-
-  return success();
-}
-
-static Type getRandomNormalLikeOutputElementType(ONNXRandomNormalLikeOp op) {
-  auto inputType = mlir::cast<TensorType>(op.getInput().getType());
-  Type elementType = inputType.getElementType();
-  if (op.getDtypeAttr()) {
-    auto builder = OpBuilder(op.getContext());
-    elementType = convertONNXTypeToMLIRType(
-        builder, static_cast<onnx::TensorProto_DataType>(
-                     op.getDtypeAttr().getValue().getSExtValue()));
-  }
-  return elementType;
+  return verifyElementTypeFromDtypeWithFallBackToInputType(*this);
 }
 
 //===----------------------------------------------------------------------===//
-// Type Inference
+// Shape + Type Inference
 //===----------------------------------------------------------------------===//
 
-std::vector<Type> ONNXRandomNormalLikeOp::resultTypeInference() {
-  Type elementType = getRandomNormalLikeOutputElementType(*this);
-  std::vector<Type> resultTypes;
-  if (auto rankedInputType =
-          mlir::dyn_cast<RankedTensorType>(getInput().getType())) {
-    resultTypes.push_back(rankedInputType.clone(elementType));
-  } else {
-    resultTypes.push_back(UnrankedTensorType::get(elementType));
-  }
-  return resultTypes;
-}
-
-//===----------------------------------------------------------------------===//
-// Shape Inference
-//===----------------------------------------------------------------------===//
-
-LogicalResult ONNXRandomNormalLikeOp::inferShapes(
-    std::function<void(Region &)> doShapeInference) {
-  if (!hasShapeAndRank(getInput()))
-    return success();
-  Type elementType = getRandomNormalLikeOutputElementType(*this);
-  return inferShapeForUnaryOps(getOperation(), elementType);
-}
+GET_SHAPE_AND_TYPE_INFERENCE_FOR_SHAPE_COPYING_OPS(ONNXRandomNormalLikeOp)
```

Comment:
@jorickert We had a discussion among us and to the extend possible, we would prefer avoiding macros for code. Would you mind reverting to having a normal definition of the 2 function calls, with inside a templated call. That way, we get the same benefits of reusing the same code (great contribution, thanks), and at the same time preserve the debug features of normal code.

Let me know if you feel strongly about this.